src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -602,8 +603,62 @@ namespace QgsWms
         }
       }
 
+      QList<QgsMapLayer *> renderLayers, auxHighlightLayers;
+
       if ( !map->keepLayerSet() )
       {
+
+        static const QRegularExpression RE_HIGHLIGHT_LAYER = QRegularExpression( "^highlight_\\d+$" );
+
+        QString presetName = map->followVisibilityPresetName();
+
+        QString newPresetName;
+        if ( !presetName.isEmpty() )
+        {
+          // get layers from theme
+          if ( c->project()->mapThemeCollection()->hasMapTheme( presetName ) )
+            renderLayers = c->project()->mapThemeCollection()->mapThemeVisibleLayers( presetName );
+          else
+            renderLayers = c->project()->mapThemeCollection()->masterVisibleLayers();
+
+          // get highlight layers, if any
+          bool has_highlight_layer = false;
+          for ( auto layer : mapSettings.layers() )
```

Comment:
```suggestion
          const QList<QgsMapLayer *> layers = mapSettings.layers();
          for ( QgsMapLayer *layer : layers )
```

NEVER iterate with a for loop directly on a list retrieved through a method (it's a silly thing of Qt ... ).
Also be explicit about return values (unless they are incredibly hard to write).